### PR TITLE
Kuz 257 authorization

### PR DESCRIPTION
**Pull request by benoitvidis**
First authorization layer implementation.

Missing: 
* Some functional tests
* Some documentation
* The permissions definition only allows some boolean value for the time being. Using a closure still to be implemented.

NB: This PR introduces a new package dependency. The container image will need to be updated once it is merged.

### CODE REVIEW

File: lib/api/controllers/funnelController.js [hunk 1]
```diff
@@ -45,7 +47,13 @@ module.exports = function FunnelController (kuzzle) {
           return false;
         }
 
-        return this[requestObject.controller][requestObject.action](requestObject, connection);
+        // check if the current user is allowed to process
+        if (context.user.profile.isActionAllowed(requestObject, context) === false) {
+          deferred.reject(new ForbiddenError('Unauthorized action [' + requestObject.collection + '/' + requestObject.controller + '/' + requestObject.action + '] for user ' + context.user._id));
+          return false;
```
Comment (scottinet) [inline lib/api/controllers/funnelController.js]: should return a new rejected promise instead of ``false`` to avoid breaking the call chain and to trigger a statistics.failedRequest() event

File: lib/api/core/models/repositories/README.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
+# Kuzzle repositories
+
+The Repository classes allow to fetch and store some Kuzlle internal business object from/to the persistent layer.
```
Comment (scottinet) [inline lib/api/core/models/repositories/README.md]: typo: Kuzlle => Kuzzle

File: lib/api/core/models/repositories/README.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+An abstract Repository class is available to be overriden if needed depending on the business needs.
+The class handles the persitence both to the CacheEngine (redis by default) and to the database (elastic search by default).
```
Comment (scottinet) [inline lib/api/core/models/repositories/README.md]: typo: persitence => persistence

File: lib/api/core/models/repositories/repository.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+Repository.prototype.loadOneFromDatabase = function (id) {
+  var
+    deferred = q.defer(),
+    requestObject,
+    result;
+
+  requestObject = new RequestObject({
+    controller: 'read',
+    action: 'get',
+    requestId: 'foo',
```
Comment (scottinet) [inline lib/api/core/models/repositories/repository.js]: should be removed to let the RequestObject constructor generate a random requestId

File: lib/api/core/models/repositories/profileRepository.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            p = new Profile();
+
+          if (result) {
+            // we got a profile from the database, we can use it
+            deferred.resolve(result);
+          }
+          else if (kuzzle.config.defaultUserProfiles[profileKey]) {
+            // no profile found in db but we have a default config for it
+            data = kuzzle.config.defaultUserProfiles[profileKey];
+            deferred.resolve(this.hydrate(p, data));
```
Comment (scottinet) [inline lib/api/core/models/repositories/profileRepository.js]: Since hydrate() returns a promise, shouldn't this be something along these lines?

```javascript
this.hydrate(p, data)
  .then(instance => deferred.resolve(instance)
  .catch(error => deferred.reject(error));
```

File: lib/api/core/models/repositories/repository.js [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+Repository.prototype.loadMultiFromDatabase = function (ids) {
+  var
+    deferred = q.defer(),
+    requestObject,
+    promises;
+
+  requestObject = new RequestObject({
+    controller: 'read',
+    action: 'mget',
+    requestId: 'foo',
```
Comment (scottinet) [inline lib/api/core/models/repositories/repository.js]: same here

File: lib/api/core/models/repositories/repository.js [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+ * Persists the given object in the collection that is attached to the repository.
+ *
+ * @param {ObjectConstructor} object The object to persist
+ * @returns {Promise}
+ */
+Repository.prototype.persistToDB = function (object) {
+  var requestObject = new RequestObject({
+    controller: 'write',
+    action: 'createOrUpdate',
+    requestId: 'foo',
```
Comment (scottinet) [inline lib/api/core/models/repositories/repository.js]: ditto

File: lib/services/elasticsearch.js [hunk 8]
```diff
@@ -81,6 +81,29 @@ module.exports = function (kuzzle, options) {
   };
 
   /**
+   * Return the list of documents matching the ids given in the body param
+   * NB: Due to internal Kuzzle mechanism, can only be called on a single index/collection,
+   * using the body { ids: [.. } syntax.
+   * @param {RequestObject} requestObject
+   * @returns {Promise}
+   */
+  this.mget = function (requestObject) {
```
Comment (scottinet) [inline lib/services/elasticsearch.js]: Nice!
I've logged a new jira ticket (KUZ-298) to implement a new API route using this method.

File: lib/api/controllers/funnelController.js (hunk 1 above)
Comment (AnthonySendra) [inline lib/api/controllers/funnelController.js]: +1 (same problem on line 47 ?)

Muste have something like:
```
deferred.reject(...);
return deferred.promise;
```

File: lib/api/controllers/writeController.js [hunk 9]
```diff
@@ -31,6 +31,9 @@ module.exports = function WriteController (kuzzle) {
         if (rooms.length > 0) {
           kuzzle.notifier.notify(rooms, new ResponseObject(requestObject).toJson());
         }
+        else if (!requestObject.isPersistent()) {
+          kuzzle.notifier.notify({}, new ResponseObject(requestObject), context.connection);
+        }
```
Comment (ballinette) [inline lib/api/controllers/writeController.js]: Why that ?
it seems not related to authorization checking.... 
Was there a bug with kuzzle response there ?

File: lib/api/core/models/repositories/README.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+* *string* id
+
+The object id to load from the cache.
+
+* *Object* opts
+
+An optional options object.
+
+Currently, the only optional parameter that can be pass to the method id *key*. If no key is given to the method, the key used by the method defaults to *collection* + '/' + id.
```
Comment (ballinette) [inline lib/api/core/models/repositories/README.md]: typo : "the only optional parameter that can be pass to the method IS *key*" (instead of "id key") ??

File: lib/api/controllers/writeController.js (hunk 9 above)
Comment (benoitvidis) [inline lib/api/controllers/writeController.js]: Indeed, this is not related to the authorization. 

This was added to cover the case where a REST non-persistent write call would not give any response but I am not sure the current implementation is the proper one, especially regarding how it will react with a webSocket connection.

Additionally, it may indeed be worth moving it into another PR.

I am checking how it can be improved/fixed.

File: lib/api/core/models/repositories/repository.js (hunk 4 above)
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: +1

File: lib/api/controllers/writeController.js (hunk 9 above)
Comment (ballinette) [inline lib/api/controllers/writeController.js]: YEp, I suggest to remove it from current PR, and create a new issue to solve that ;)

**Comment by ballinette** [comment]
Oh my god, I have an erection !!!
PR OK for me, despite nitpicking remarks above ;)

File: lib/api/core/models/repositories/repository.js [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+    controller: 'read',
+    action: 'mget',
+    requestId: 'foo',
+    collection: this.collection,
+    body: {
+      ids: ids
+    }
+  });
+
+  this.readEngine.mget(requestObject)
```
Comment (AnthonySendra) [inline lib/api/core/models/repositories/repository.js]: Sorry, I'm lost :). Where the index is defined ?
Because in elasticsearch service we force the mainindex in cleanData.
Or it's not on specific index and just on specific collection ?

**Comment by AnthonySendra** [comment]
+1000

**Comment by EmilieEsposito** [comment]
Jeff, get out of Gilles'body !

On Thu, Nov 5, 2015 at 6:27 PM, Gilles Ballini <notifications@github.com>
wrote:

> Oh my god, I have an erection !!!
> PR OK for me, despite nitpicking remarks above ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kuzzleio/kuzzle/pull/63#issuecomment-154130447>.
>



-- 


*Emilie Esposito - Kuzzle Team Product Owner*
emilie@kuzzle.io

------------------------------
kuzzle.io
@kuzzleio <https://twitter.com/kuzzleio>
Fork it on GitHub <https://github.com/kuzzleio>

File: lib/api/core/models/repositories/repository.js (hunk 11 above)
Comment (benoitvidis) [inline lib/api/core/models/repositories/repository.js]: Indeed, for the time being, the index is not taken into account.

On the role level, you can already define some permissions on an index but it actually won't be used and you need a wildcard entry for the time being.

File: lib/api/controllers/funnelController.js (hunk 1 above)
Comment (benoitvidis) [inline lib/api/controllers/funnelController.js]: Indeed. Fixed

File: lib/api/core/models/repositories/README.md (hunk 2 above)
Comment (benoitvidis) [inline lib/api/core/models/repositories/README.md]: fixed

File: lib/api/core/models/repositories/README.md (hunk 3 above)
Comment (benoitvidis) [inline lib/api/core/models/repositories/README.md]: fixed

File: lib/api/core/models/repositories/repository.js (hunk 4 above)
Comment (benoitvidis) [inline lib/api/core/models/repositories/repository.js]: fixed

File: lib/api/core/models/repositories/repository.js (hunk 6 above)
Comment (benoitvidis) [inline lib/api/core/models/repositories/repository.js]: fixed
